File: week2/vigenere.c
// Dziugas Mikalkenas
// 29 March 2019
// Problem Set 2

#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>

int isInputAlpha(string prompt);
int shift(char c);
void printCiphered(unsigned char ciphered_char, int higherThan);

int main(int argc, string argv[])
{
    if (argc == 2)
    {        
        isInputAlpha(argv[1]);        
        string plaintext = get_string("plaintext: ");

        int plaintextLength = strlen(plaintext);
        int keywordLength = strlen(argv[1]);
        
        if (plaintextLength == 0)
        {
            printf("Please specify plaintext\n");
            return 1;
        }

        int i = 0;
        int j = 0;
        
        printf("ciphertext: ");
        
        while (i < keywordLength)
        {
            int key = shift(argv[1][i]);
            //printf("current key: %i\n", key);
            while (j < plaintextLength)
            {
                if (isalpha(plaintext[j]))
                {
                    //printf("plainchar: %i \n", plaintext[j]);
                    unsigned char ciphered_char = plaintext[j] + key % 26;
                    //printf("cipherchar: %i \n", ciphered_char);
                    if (islower(plaintext[j]))
                    {
                        printCiphered(ciphered_char,122);
                    }
                    else if (isupper(plaintext[j]))
                    {
                        printCiphered(ciphered_char,90);
                    }
                    i++;
                    j++;
                    break;
                }
                else
                {
                    printf("%c", plaintext[j]);
                    j++;
                }               
            }
            if (i == keywordLength)
            {
                i = 0;
            }
            if (j == plaintextLength)
            {
                break;
            }
        }
        printf("\n");
    }
    else
    {
        printf("Usage: ./vigenere <keyword>\n");
        return 1;
    }
}

int isInputAlpha(string prompt)
{
    int stringLength = strlen(prompt);
    for (int i = 0; i < stringLength; i++)
    {
        if (!isalpha(prompt[i]))
        {
            printf("Usage: ./vigenere <keyword>\n");
            exit(1);
        }
    }
    return atoi(prompt);
}

int shift(char c)
{
    if (isupper(c))
    {
        return c - 65;
    }
    else if (islower(c))
    {
        return c - 97;
    }
    else
    {
        exit(1);
    }
}

void printCiphered(unsigned char ciphered_char, int higherThan)
{
    if (ciphered_char > higherThan)
    {
        printf("%c", ciphered_char - 26);
        //printf("%i\n", ciphered_char - 26);
    }
    else
    {
        printf("%c", ciphered_char);
        //printf("%i\n", ciphered_char);                          
    } 
}

File: week6/vigenere.py
import sys
import string
from cs50 import get_string

# Checks for corrent number of arguments
if len(sys.argv) != 2:
    print("Usage: python vigenere.py <keyword>")
    exit(1)

# Checks if keyword is alphabetic
if not sys.argv[1].isalpha():
    print("Usage: python vigenere.py <keyword>")
    exit(1)

k = sys.argv[1].lower()
keywordLength = len(k)
p = get_string("plaintext: ")
plaintextLength = len(p)

ciphertext = ""

i = 0
j = 0

# Enciphers
while i < keywordLength:
    while j < plaintextLength:
        if p[j] in string.ascii_uppercase:
            ciphered_char = ((string.ascii_uppercase.index(p[j]) + string.ascii_lowercase.index(k[i])) % 26) + 65
            ciphertext += chr(ciphered_char)
            i += 1
            j += 1
        elif p[j] in string.ascii_lowercase:
            ciphered_char = ((string.ascii_lowercase.index(p[j]) + string.ascii_lowercase.index(k[i])) % 26) + 97
            ciphertext += chr(ciphered_char)
            i += 1
            j += 1
        else:
            ciphertext += p[j]
            j += 1

        if i == keywordLength:
            i = 0
    if j == plaintextLength:
        break

print(f'ciphertext: {ciphertext}') 
